r2i/onnxrt/engine.cc

```diff
@@ -0,0 +1,331 @@
+ * back to RidgeRun without any encumbrance.
+ */
+
+#include "r2i/onnxrt/engine.h"
+
+#include <core/common/exceptions.h>
+#include <core/session/onnxruntime_cxx_api.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "r2i/onnxrt/frame.h"
+#include "r2i/onnxrt/model.h"
+#include "r2i/onnxrt/prediction.h"
+
+namespace r2i {
+namespace onnxrt {
+
+Engine::Engine () : state(State::STOPPED), model(nullptr) {
+}
+
+Engine::~Engine () {
+  this->Stop();
+}
+
+RuntimeError Engine::SetModel (std::shared_ptr<r2i::IModel> in_model) {
+  RuntimeError error;
+
+  if (State::STOPPED != this->state) {
+    error.Set (RuntimeError::Code::WRONG_ENGINE_STATE,
+               "Stopped model before setting a new state");
+    return error;
+  }
+
+  if (nullptr == in_model) {
+    error.Set (RuntimeError::Code:: NULL_PARAMETER,
+               "Received null model pointer");
+    return error;
+  }
+  auto model = std::dynamic_pointer_cast<r2i::onnxrt::Model, r2i::IModel>
+               (in_model);
+
+  if (nullptr == model) {
+    error.Set (RuntimeError::Code::FRAMEWORK_ERROR,
+               "The provided model is not an ONNXRT model");
+    return error;
+  }
+
+  this->model = model;
+
+  return error;
+}
+
+RuntimeError Engine::Start ()  {
+  RuntimeError error;
+
+  if (State::STARTED == this->state) {
+    error.Set (RuntimeError::Code::WRONG_ENGINE_STATE,
+               "Engine already started");
+    return error;
+  }
+
+  if (nullptr == this->model) {
+    error.Set (RuntimeError::Code:: NULL_PARAMETER,
+               "Model not set yet");
+    return error;
+  }
+
+  try {
+    this->num_input_nodes = this->GetSessionInputCount(
+                              this->model->GetOnnxrtSession());
+    this->num_output_nodes = this->GetSessionOutputCount(
+                               this->model->GetOnnxrtSession());
+  }
+
+  catch (std::exception &excep) {
+    error.Set(RuntimeError::Code::FRAMEWORK_ERROR, excep.what());
+    return error;
+  }
+
+  if (this->num_input_nodes > 1) {
+    error.Set(RuntimeError::Code::INCOMPATIBLE_MODEL,
+              "Number of inputs in the model is greater than 1, this is not supported");
+    return error;
+  }
+
+  if (this->num_output_nodes > 1) {
+    error.Set(RuntimeError::Code::INCOMPATIBLE_MODEL,
+              "Number of outputs in the model is greater than 1, this is not supported");
+    return error;
+  }
+
+  this->input_node_names.resize(num_input_nodes);
+  this->output_node_names.resize(num_output_nodes);
+
+  // Warning: only 1 input and output supported
+  error = GetSessionInfo(this->model->GetOnnxrtSession(), 0);
+  if (error.IsError ()) {
+    return error;
+  }
+
+  this->state = State::STARTED;
+
+  return error;
+}
+
+RuntimeError Engine::Stop () {
+  RuntimeError error;
+
+  if (State::STOPPED == this->state) {
+    error.Set (RuntimeError::Code::WRONG_ENGINE_STATE,
+               "Engine already stopped");
+  }
+
```

Comment:
Don't we have to do something with the ONNX API here?

Reply:
I don't believe there is.